### PR TITLE
Learning ebpf_exporter by adding examples/cgroup-rstat-flushing

### PR DESCRIPTION
**Pull request by netoptimizer**
This PR implements example cgroup-rstat-flushing

- For analyzing cgroup rstat flushing behavior and lock contention issue observed in production

Thanks to for @bobrik for reviewing as I need to learn how to use ebpf_exporter

- For other to learn, I've kept some commits that show mistakes and followup commits that fixes these

### CODE REVIEW

File: examples/cgroup-rstat-flushing.bpf.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+ Concern: The lock rates will vary (a lot) and aggregating this as an average
+ will not capture spikes. Especially given Prometheus capture intervals only
+ happens every 53 seconds.
+
+ Q: will lock "type=yield" be a good idea?
+
+*/
+
+SEC("tp_btf/cgroup_rstat_locked")
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:72]: This is too new for the Ubuntu we run in CI, so you need to add an exclude to `CONFIGS_TO_IGNORE_IN_CHECK` in the top level `Makefile`.

File: examples/cgroup-rstat-flushing.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+          decoders: # contended boolean converted to 0 and 1
+            - name: uint
+    - name: cgroup_rstat_lock_contended
+      help: Lock contention counters per cgroup level
+      labels:
+        - name: level
+          size: 4
+          decoders:
+            - name: uint
+
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.yaml:31]: Let's drop this empty line.

File: examples/cgroup-rstat-flushing.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+          decoders:
+            - name: uint
+    - name: cgroup_rstat_locked_total
+      help: Total number of times rstat lock was obtainted and contended state
+      labels:
+        - name: contended
+          size: 4
+          decoders: # contended boolean converted to 0 and 1
+            - name: uint
+    - name: cgroup_rstat_locked_yield
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.yaml:17]: `cgroup_rstat_locked_yields_total`, see https://prometheus.io/docs/practices/naming/

File: examples/cgroup-rstat-flushing.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      help: Number of times rstat lock was obtainted again after yield and contended state
+      labels:
+        - name: contended
+          size: 4
+          decoders: # contended boolean converted to 0 and 1
+            - name: uint
+    - name: cgroup_rstat_lock_contended
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.yaml:24]: `cgroup_rstat_lock_contentions_total`, I think.

File: examples/cgroup-rstat-flushing.bpf.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+ */
+
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "maps.bpf.h"
+
+#define MAX_CGRP_LEVELS	5
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:17]: The tabs are making clang-format unhappy in CI.

File: examples/cgroup-rstat-flushing.bpf.c [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+ Mitigation workaround: Store counters per cgroup "level" (level=0 is root).
+ This will allow us to separate root-cgroup flushes from cadvisor walking all
+ cgroup levels.
+
+ */
+SEC("fentry/cgroup_rstat_flush_locked")
+int BPF_PROG(cgroup_rstat_flush_locked, struct cgroup *cgrp)
+{
+	u64 level_key = cgrp->level;
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:129]: Should this be u32 to match `__type(key, u32)`?

File: examples/cgroup-rstat-flushing.bpf.c [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+*/
+
+SEC("tp_btf/cgroup_rstat_locked")
+int BPF_PROG(rstat_locked, struct cgroup *cgrp, int cpu, bool contended)
+{
+	u32 key = contended;
+	u64 *cnt;
+
+	read_array_ptr(&cgroup_rstat_locked_total, &key, cnt);
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:78]: Shouldn't `increment_map_nosync` work here as well?

File: examples/cgroup-rstat-flushing.bpf.c [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+ *  - https://github.com/xdp-project/xdp-project/blob/master/areas/latency/cgroup_rstat_tracepoint.bt
+ *
+ * Depends on tracepoints added in kernel v6.10
+ */
+
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "maps.bpf.h"
+
+#define MAX_CGRP_LEVELS	5
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:14]: nit: `MAX_CGROUP_LEVELS`

File: examples/cgroup-rstat-flushing.bpf.c [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+	read_array_ptr(&cgroup_rstat_locked_total, &key, cnt);
+	(*cnt)++;
+
+	if (cpu >= 0) {
+		read_array_ptr(&cgroup_rstat_locked_yield, &key, cnt);
+		(*cnt)++;
+	}
+
+	/* What cgrp level is interesting, but I didn't manage to encode it in
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:86]: nit: `cgrp` -> `cgroup`

File: examples/cgroup-rstat-flushing.bpf.c (hunk 5 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:17]: IMHO then BPF  restricted-C code should be written with same tabs style as kernel code.
Lets change for formatting requirements, please!

File: examples/cgroup-rstat-flushing.bpf.c (hunk 1 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:72]: Makes sense.
I'll do that.

File: examples/cgroup-rstat-flushing.yaml (hunk 2 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.yaml]: done

File: examples/cgroup-rstat-flushing.bpf.c (hunk 6 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:129]: This is a good catch - thanks!

File: examples/cgroup-rstat-flushing.bpf.c (hunk 8 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:14]: sure

File: examples/cgroup-rstat-flushing.bpf.c (hunk 9 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:86]: Sure to nit.
I was hoping to get your input to the contents of the comment:
```
	/* What cgrp level is interesting, but I didn't manage to encode it in
	 * above counters.  As contended case is the most interesting, have
	 * level counter for contended.
	 */
```
When coding BPF, I usually do **one** map lookup to get a stats `struct`, that I will update with multiple stats.
Here I end-up doing 3 map lookups for single counters.

When I stated coding I though I could represent something like:
```
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="0"} 6558
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="0"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="1"} 34
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="1"} 123
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="5"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="5"} 0
```
Question: But ...
1. I cannot figure out how to get ebpf_exporter to generate this?
2. I don't know if these counters makes sense for Prometheus ?

File: examples/cgroup-rstat-flushing.yaml (hunk 3 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.yaml:17]: Thanks for the [link](https://prometheus.io/docs/practices/naming/) , as a Prometheus newbie I didn't realize that the __suffix_ had a meaning.
I now realize that general strucure is _library_name_unit_suffix_.

Jakub recommended: [Prometheus: Up & Running, 2nd Edition](https://www.oreilly.com/library/view/prometheus-up/9781098131135/)

File: examples/cgroup-rstat-flushing.bpf.c (hunk 7 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:78]: Code just evolved from doing some operations on the returned pointer "object".
So, now this is a pointer to a simple counter I can use `increment_map_nosync` as I do in other places.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 9 above)
Comment (bobrik) [reply examples/cgroup-rstat-flushing.bpf.c]: > I don't know if these counters makes sense for Prometheus ?

They do

> I cannot figure out how to get ebpf_exporter to generate this?

Each line here is a separate key in a single map called `ebpf_exporter_cgroup_rstat_locked_total`:

```
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="0"} 6558
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="0"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="1"} 34
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="1"} 123
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="5"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="5"} 0
```

The mapping is effectively:

* `{contended="0", level="0"}` -> `6558`
* `{contended="1", level="0"}` -> `0`
* `{contended="0", level="1"}` -> `34`
* ...

To encode each key, you turn the raw bytes of the key into the values with the help of decoders.

Working backwards from what you expect on the output, you want something like this in the config:

```yaml
    - name: cgroup_rstat_locked_total
      help: Total number of times rstat lock was obtainted and contended state
      labels:
        - name: contended
          size: 1
          decoders:
            - name: uint
        - name: level
           size: 1
           decoders:
            - name: uint
```

On the kernel side you'd want something like this to represent the key:

```c
struct key_t {
    u8 contended;
    u8 level;
};
```

The names here do not matter, only sizes do (u8 corresponds to `size: 1` in the config).

Given that your keys are complex structs and not small numbers, you cannot store these in an array, so you need a hashmap:

```
struct {
    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
    __uint(max_entries, 128);
    __type(key, struct key_t);
    __type(value, u64);
} cgroup_rstat_locked_total SEC(".maps");
```

To actually increment a value, you need to construct the key and to the increment in the map:

```c
// construct however you need to
struct key_t key = {};
key.contended = 1;
key.level = 2;

increment_map_nosync(&cgroup_rstat_locked_total, &key, 1);
```

Here's an example with a complex key:

* https://github.com/cloudflare/ebpf_exporter/blob/master/examples/pci.yaml
* https://github.com/cloudflare/ebpf_exporter/blob/master/examples/pci.bpf.c

Hope this makes sense.

Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c]: These suggestions worked for encoding more information ("labels" in Prometheus terms). 
I've coded it up and pushed as part of this patchset/PR.
(p.s. I've also encoded "yield" state)

File: examples/cgroup-rstat-flushing.bpf.c (hunk 5 above)
Comment (bobrik) [reply examples/cgroup-rstat-flushing.bpf.c:17]: Let's leave it for another issue. For now let's just make clang-format happy.

File: examples/cgroup-rstat-flushing.bpf.c [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+	u64 now = bpf_ktime_get_ns();
+	u64 pid = bpf_get_current_pid_tgid();
+	u64 *start_hold_ts;
+	struct hist_key_t key;
+	u64 delta_usec;
+
+	/* Lock hold time */
+	read_array_ptr(&start_hold, &pid, start_hold_ts);
+	// TODO: validate LRU lookup success
+	delta_usec = (now - *start_hold_ts) / 100; /* 0.1 usec */
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:154]: This isn't really `usec` anymore. Any reason not to keep it in ns?

File: examples/cgroup-rstat-flushing.bpf.c [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+	u32 level = cgrp->level;
+
+	if (level > MAX_CGROUP_LEVELS)
+		level = MAX_CGROUP_LEVELS;
+
+	if (cpu >= 0)
+		lock_key.yield = 1;
+
+	lock_key.contended = contended;
+	lock_key.level = (level & 0xFF);
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:113]: What does this do?

File: examples/cgroup-rstat-flushing.bpf.c [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+ happens every 53 seconds.
+
+*/
+SEC("tp_btf/cgroup_rstat_locked")
+int BPF_PROG(rstat_locked, struct cgroup *cgrp, int cpu, bool contended)
+{
+	u64 now = bpf_ktime_get_ns();
+	u64 pid = bpf_get_current_pid_tgid();
+	struct lock_key_t lock_key = { 0 };
+	u32 level = cgrp->level;
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:104]: You can operate on `lock_key.level` directly, no need for a separate variable.

File: examples/cgroup-rstat-flushing.bpf.c [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+	increment_map_nosync(&cgroup_rstat_locked_total, &lock_key, 1);
+
+	/* Lock hold time start */
+	bpf_map_update_elem(&start_hold, &pid, &now, BPF_ANY);
+	// TODO: Should we ignore yield and measure flush time instead?
+
+	/* Lock contended event happened prior to obtaining this lock.
+	 * Get back start "wait" timestamp that was recorded.
+	 */
+	if (contended) {
```
Comment (bobrik) [inline examples/cgroup-rstat-flushing.bpf.c:124]: This can be an early return.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 10 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:154]: Good point with variable `usec` naming, will fix.

Keeping it in nanosec will use more buckets, and I'm trying to reduce how much data we feed Prometheus.
Maybe I don't need to reduce size?

File: examples/cgroup-rstat-flushing.bpf.c (hunk 13 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:124]: True for current code, but I might add more code later.
And I think the code is more readable IMHO, as it is more clear that everything under `if (contended) {` is handling the contended case.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 12 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:104]: Below I'm using `level` and bounding it to `MAX_CGROUP_LEVELS`.
I prefer to keep level type `u32` in this bounding code, and do the assignment later to `lock_key.level` as it is type `u16`.

File: examples/cgroup-rstat-flushing.yaml (hunk 4 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.yaml]: This map was removed after improvement to encode this as labels.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 5 above)
Comment (netoptimizer) [reply examples/cgroup-rstat-flushing.bpf.c:17]: This projects `.clang-format` file states:
`# Minimal format loosely matching the kernel style.`
IMHO it doesn't match kernel style at all...

As it contains:
```
IndentWidth: 4
TabWidth: 4
UseTab: Never
```
Looking at kernel source in `.clang-format` this should have been:
```
IndentWidth: 8
TabWidth: 8
UseTab: Always
```

I will respect that this project requires its own special clang-format rule and obey to these, but I still think they are wrong.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 10 above)
Comment (bobrik) [reply examples/cgroup-rstat-flushing.bpf.c]: I personally would stick to a usec level resolution. Not sure if you care about buckets below 1us.

Whatever works for you is fine, as long as the naming doesn't contradict the contents. We can always change it later.

File: examples/cgroup-rstat-flushing.bpf.c (hunk 11 above)
Comment (bobrik) [reply examples/cgroup-rstat-flushing.bpf.c]: I would normally just cast it explicitly: `lock_key.level = (u16) level`, which should have the same effect.

**Comment by bobrik** [comment]
Let me know when it's ready to go. It would be helpful if you attached the exact metrics that you see locally with this running and capturing some events.

**Comment by netoptimizer** [comment]
Running this locally overnight, I see an issue with cgroup decoding: e.g `cgroup="unknown_cgroup_id:2196814`
Longer output:
```
ebpf_exporter_cgroup_rstat_flush_seconds_sum{cgroup="unknown_cgroup_id:2196562",level="2"} 224
ebpf_exporter_cgroup_rstat_flush_seconds_sum{cgroup="unknown_cgroup_id:2196598",level="2"} 305
ebpf_exporter_cgroup_rstat_flush_seconds_sum{cgroup="unknown_cgroup_id:2196670",level="2"} 172
ebpf_exporter_cgroup_rstat_flush_seconds_sum{cgroup="unknown_cgroup_id:2196706",level="2"} 167
ebpf_exporter_cgroup_rstat_flush_seconds_sum{cgroup="unknown_cgroup_id:2196742",level="2"} 210
[...]
curl -s localhost:9435/metrics | grep unknown_cgroup_id | grep _sum | wc -l
6043
```
Something on my system seems to be creating cgroups for a short period and ebpf_exporter is picking these up
-  I assume what happens is that:
- when decoding these for (Prometheus) output, they are gone and we get these "unknown_cgroup_id"
- (for completeness: I think `pmcd` / `pmlogger` is causing this, but it is irrelevant)

I'm want to change the code such that this per-cgroup stats cannot explode like this
- considering using a LRU map with a limit (e.g. 1024)
- as we are primarily interested in the most active cgroups
- but as `cadvisor` walks all cgroups we have to be careful limit is always above cgroups in system to avoid LRU trashing corner-cases

**Comment by netoptimizer** [comment]
Regarding above "unknown_cgroup_id:nnn" issue (Cc @bobrik )

Can the _cgroup_ **decoder** be instructed to exclude unknown cgroups in the output?
- or only make a single "unknown_cgroup" output record?

**Comment by bobrik** [comment]
> Something on my system seems to be creating cgroups for a short period and ebpf_exporter is picking these up

Does ebpf_exporter say anything about `fanotify` (#244) on startup? Even short lived cgroups should be resolved when fanotify is available.

> Can the cgroup decoder be instructed to exclude unknown cgroups in the output?
> or only make a single "unknown_cgroup" output record?

Not in the current state, but it seems like a good idea: #460.

Another option, which might not be possible: check for cgroup age and if it's less than some threshold (say 10s), set cgroup_id to zero to do the collapsing manually.

**Comment by netoptimizer** [comment]
> > Something on my system seems to be creating cgroups for a short period and ebpf_exporter is picking these up
> 
> Does ebpf_exporter say anything about `fanotify` (#244) on startup? Even short lived cgroups should be resolved when fanotify is available.
> 

I don't think `fanotify` is avail/configured on my testlab.

> > Can the cgroup decoder be instructed to exclude unknown cgroups in the output?
> > or only make a single "unknown_cgroup" output record?
> 
> Not in the current state, but it seems like a good idea: #460.

Awesome that you created a ticket for this :smiley:

**Comment by bobrik** [comment]
> I don't think fanotify is avail/configured on my testlab.

It's a matter of having v6.6 or newer: https://github.com/torvalds/linux/commit/0ce7c12e88cf. There isn't anything to configure. As long as ebpf_exporter doesn't say anything about fanotify at startup, it is using it.

**Comment by netoptimizer** [comment]
Some example output running this in "localhost" mode
- e.g. on port 9435
-  extracted via command: `curl -s localhost:9435/metrics`

```
# HELP ebpf_exporter_cgroup_rstat_locked_total Times rstat lock was obtainted with state for cgroup level, contended and yield
# TYPE ebpf_exporter_cgroup_rstat_locked_total counter
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="0",yield="0"} 24281
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="0",yield="1"} 7351
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="1",yield="0"} 14533
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="1",yield="1"} 39
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="2",yield="0"} 175181
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="2",yield="1"} 89
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="3",yield="0"} 272801
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="3",yield="1"} 119
ebpf_exporter_cgroup_rstat_locked_total{contended="0",level="4",yield="0"} 1805
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="0",yield="0"} 86951
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="0",yield="1"} 1.2791901e+07
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="1",yield="0"} 73
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="1",yield="1"} 3864
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="2",yield="0"} 722
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="2",yield="1"} 22764
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="3",yield="0"} 1077
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="3",yield="1"} 34344
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="4",yield="0"} 10
ebpf_exporter_cgroup_rstat_locked_total{contended="1",level="4",yield="1"} 399
```

**Review by bobrik — APPROVED** [review]

**Pull request merged**